### PR TITLE
feat: enforce CRC ICT consistency on write

### PR DESCRIPTION
**Pull request by scottsand-db**
## What changes are proposed in this pull request?

It's _possible_ to load a Snapshot with slightly invalid state: somehow the latest commit was _missing_ an ICT despite the table feature being enabled.

For now, Kernel only fails when we _ask_ for the timestamp of such a Snapshot.

Soon, we will be able to incrementally load a CRC at that Snapshot version, too.

This PR prevents us from _writing_ out that in-memory CRC at that Snapshot version, which would yield an invalid CRC!

## How was this change tested?

Simple new tests.

### CODE REVIEW

File: kernel/src/crc/writer.rs [hunk 1]
```diff
@@ -22,6 +25,19 @@ pub(crate) fn try_write_crc_file(engine: &dyn Engine, path: &Url, crc: &Crc) ->
             crc.file_stats_state
         ))
     );
+    let ict_enabled = crc
+        .metadata
+        .parse_table_properties()
```
Comment (kyli87) [inline kernel/src/crc/writer.rs]: Seems excessive to have to parse the entire set of TableProperties just to check one field. You could consider exposing a single property getter on `Metadata` or use `Metadata.configuration()` and directly check the single string `ENABLE_IN_COMMIT_TIMESTAMPS` key of the hashmap?

File: kernel/src/crc/writer.rs [hunk 2]
```diff
@@ -55,8 +80,13 @@ mod tests {
             ],
         )
         .unwrap();
+        let metadata = if ict_enabled {
+            Metadata::default().with_configuration_entry("delta.enableInCommitTimestamps", "true")
```
Comment (kyli87) [inline kernel/src/crc/writer.rs]: nit: one additional edge case to test is `"delta.enableInCommitTimestamps"="false"` (which I assume is also valid?)

Comment (kyli87) [inline kernel/src/crc/writer.rs]: nit: use the existing `ENABLE_IN_COMMIT_TIMESTAMPS` constant

**Review by kyli87 — APPROVED** [review]
LGTM - some minor comments/suggestions

**Review by rliao147 — APPROVED** [review]
LGTM, thanks for adding the guardrails.

File: kernel/src/crc/writer.rs (hunk 1 above)
Comment (scottsand-db) [reply kernel/src/crc/writer.rs]: > Seems excessive to have to parse the entire set of TableProperties just to check one field.

+1, but that's the setup we have today. Kernel-Java let's you cleanly check a simple property using nice type defs against the metadata hashmap. Kernel Rust does not -- you have to parse the entire thing.

cc @nicklan can we revisit this structure?

> You could consider exposing a single property getter on Metadata

Yes but we already have similar APIs on TableConfigruation and on Protocol. IMO our whole Metadata + Protocol + TableConfiguration abstraction is a mess.

> use Metadata.configuration() and directly check the single string ENABLE_IN_COMMIT_TIMESTAMPS key of the hashmap?

Will do this for now

Comment (nicklan) [reply kernel/src/crc/writer.rs]: Yeah, I'm not wedded to our current abstractions at all. If you want to propose a new way of doing things, feel free :)
